_posts/en/newsletters/2020-06-10-newsletter.md

```diff
@@ -0,0 +1,311 @@
+  those UTXOs.  The spending transaction simply commits to the txid and
+  location of the UTXOs it wants to spend, requiring other software that
+  wants to calculate the fee to either look up each UTXO's previous
+  transaction or to maintain a database of verified UTXO data.
+
+    Hardware wallets don't maintain a UTXO set, so the only trustless
+    way for them to determine the amount of fee paid by a transaction
+    with legacy UTXOs is to get a copy of each legacy UTXO's previous
+    transaction, hash the previous transaction to ensure its txid
+    matches the UTXO reference, and use the now-verified amount of the
+    UTXO to perform the fee calculation.  Since legacy transactions can
+    be almost as large as one megabyte and it's possible for a spending
+    transaction to reference thousands of previous transactions, this
+    process can require the processing of gigabytes of data by
+    resource-constrained hardware wallets.
+
+    One of the several improvements made in [BIP143][] segwit v0 was an
+    attempt to eliminate this burden by having [signatures commit to the
+    amount][bip143 motivation] of the UTXO they were spending.  This
+    meant that any signatures that committed to an incorrect amount
+    would be invalid, which developers of both [Bitcoin Core][towns
+    benefits] and [hardware wallets][palatinus inpatient] believed would
+    allow signers to safely accept amounts from untrusted programs.
+
+    Unfortunately, signing a single UTXO's amount turned out to be
+    insufficient.  In 2017, Greg Sanders [described][sanders attack] how
+    an attacker could send two versions of the same transaction with two
+    segwit UTXOs to a hardware wallet.  In the first case, the attacker
+    would understate the amount of one of the UTXOs being spent; the
+    second case, they'd understate the amount of the second UTXO.
+    Because both transactions understated the amount of one of the
+    UTXOs, the hardware wallet would under-calculate the fee by that
+    same understated amount.  A user would authorize signing the first
+    transaction based on its understated fee.  The attacker would then
+    claim there was some minor problem and the user needs to sign the
+    same transaction again, but would instead get the user to sign the
+    second transaction (which would look identical to the user).
+    Although both transactions produced are individually invalid because
+    they each contain one signature that commits to a wrong amount, each
+    transaction also contains one signature that's valid.  The attacker
+    uses those two valid signatures to synthesize a valid transaction
+    that overpays its fees.
+
+    ![Fee overpayment attack illustration](/img/posts/2020-06-fee-overpayment-attack.dot.png)
+
+    Similar to the other attack described in Sander's 2017 email (see
+    [Newsletter #97][news97 spk commit]), this attack only affects
```

Comment:
OOC what was newly revealed this week vs. Sander's 2017 email? Why wasn't this discussed more and addressed 3 years ago?

Reply:
I don't know what "OOC" means, but as far as I know nothing new has been revealed this week.  I think it was discussed a bit more back then (e.g. on IRC) and this summary describes the attempts to resolve it at the protocol layer (e.g. Lau's SIGHASH2 proposal in 2018 and bip-taproot in 2019, both of which we summarized in Optech newsletters).

It appears from Trezor's description that they just learned about this attack when Rashid reported it and I'm guessing that that's also when other hardware wallets learned about it (or through Trezor's own disclosure, as it appears Rashid didn't contact Coinkite (ColdCard) or some other hardware wallet vendors).  I've been aware of this attack since shortly after Sander's published his description (I think I first heard about it on IRC) and it never occurred to me that it wasn't known to vendors.

I have a pet theory that Trezor didn't see Sanders's post because it occurred in a thread about PSBT and Trezor has historically been [opposed](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013041.html) to hardware wallet interface standardization, Ledger has never been really involved in Bitcoin protocol development, and CoinKite hadn't started work on ColdCard at that point.

As to why nobody else worked on it, comments in #bitcoin-core-dev IRC indicate that nobody who knew about the attack for the past three years thought it was particularly urgent.  I'll update to sprinkle a few more links through the text.  My own personal take on it is that it isn't that severe of an issue but that probably all hardware wallet manufacturers will end up doing the same thing Trezor does just to prevent FUD from other HW manufacturers and to cover their asses against people claiming to lose money due to this vulnerability (whether legitimate or illegitimate---if you're a miner or know a trusted miner, it's easy to fake being a victim of this attack).

Reply:
@harding Well said. This is my understanding as well, from the various IRC discussions of the past week.

Reply:
At the last wallet meeting Friday, people weren't immediately convinced it was worth patching: http://www.erisian.com.au/bitcoin-core-dev/log-2020-06-05.html#l-346